### PR TITLE
Calculate class hashes when building

### LINKED ISSUE #1163: Support calculating `class_hash` of a contract
For use cases like using `declare` transaction through external tools (e.g. starknet-jvm), it'd be usefel to be able calculate `class_hash` along with compiling a contract through Protostar. 

One possible usecase: `class_hash` is used to calculate a hash of `Declare` transactions which is then used to create a signature.

### PR DESCRIPTION
**Pull request by karol-bisztyga**
fix #1163 

I think there are two options here (at least):
- return the class hash when the `build` command is called
- create a new command that will calculate the contract's class hash

I think the former is simpler and would go that way, but I'm open to suggestions.

If we proceed with this, we may want to make outputting the class hash optional as it may not always be needed.

TODO:
- [x] consider printing the class hash optional (do it only when the user provides a flag like `--print-class-hash`)
- [x] docs
- [x] think of integration/e2e tests

### CODE REVIEW

File: protostar/commands/build_command.py [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
     def format_human(self, fmt: LogColorProvider) -> str:
-        return "Building projects' contracts"
+        lines: list[str] = ["Building projects' contracts"]
+        for contract_name, class_hash in self.response.class_hashes.items():
+            lines.append(f"Class hash for contract {contract_name}: {class_hash}")
+
+        return "\n".join(lines)
+
+    def format_dict(self) -> dict:
+        return self.response.class_hashes
```
Comment (kasperski95) [inline protostar/commands/build_command.py:24]: Class hashes should be represented as hexes preceded with `0x`.

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Good job! Can we write an integration or e2e test testing that?

File: protostar/starknet_gateway/gateway_response.py [hunk 2]
```diff
@@ -27,3 +27,8 @@ class SuccessfulDeployAccountResponse:
     address: Address
     transaction_hash: int
     code: str
+
+
+@dataclass
+class SuccessfulBuildResponse:
```
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_response.py:33]: Sorry but I'm not sure what you meant exactly. What `Summary` do you mean?

**Comment by karol-bisztyga** [comment]
> Good job! Can we write an integration or e2e test testing that?

Thanks! I added writing tests to the todo list in the description, I will take care of this along with the docs.

File: protostar/commands/build_command.py [hunk 3]
```diff
@@ -1,15 +1,27 @@
 from pathlib import Path
 from typing import List, Optional, Any
+from dataclasses import dataclass
 
 from protostar.cli import ProtostarArgument, ProtostarCommand, MessengerFactory
 from protostar.cli.common_arguments import COMPILED_CONTRACTS_DIR_ARG
 from protostar.compiler import ProjectCompiler, ProjectCompilerConfig
-from protostar.io import LogColorProvider, Message
+from protostar.starknet_gateway import SuccessfulBuildResponse
```
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/commands/build_command.py]: [move and rename SuccessfulBuildResponse](https://github.com/software-mansion/protostar/pull/1244/commits/f98d132bbffdeabed47f5e8feedba598ac8aaa16)

File: protostar/commands/build_command.py (hunk 1 above)
Comment (karol-bisztyga) [reply protostar/commands/build_command.py]: [convert class hashes to the hex format](https://github.com/software-mansion/protostar/pull/1244/commits/98751f8e3808ac887f999c58b31a3a5918f00bc0)

File: protostar/starknet_gateway/gateway_response.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/gateway_response.py]: Message 😅 , it is named `FormatSummary` for formatting

[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_response.py]: [remove unnecessary class](https://github.com/software-mansion/protostar/pull/1244/commits/12150a336b13428628ed13b789ae3026768e5252)

**Review by kasperski95 — APPROVED** [review]

File: protostar/commands/build_command.py [hunk 4]
```diff
@@ -1,15 +1,29 @@
...
-class BuildActivityMessageTemplate(Message):
+@dataclass
+class SuccessfulBuildMessage(StructuredMessage):
+    class_hashes: dict[str, int]
+
     def format_human(self, fmt: LogColorProvider) -> str:
-        return "Building projects' contracts"
+        lines: list[str] = ["Building projects' contracts"]
+        for contract_name, class_hash in self.class_hashes.items():
+            lines.append(f"Class hash for contract {contract_name}: {class_hash}")
```
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/commands/build_command.py]: yes, sorry I missed that [do hex in format human](https://github.com/software-mansion/protostar/pull/1244/commits/fed8dd1ab4542b1d05b7581790e198b97aa2454d)

**Comment by karol-bisztyga** [comment]
I'm not sure whether an integration test is necessary here. The change impacts only what users see I don't see a change in how the parts of the application work together. If you think otherwise, please, let me know.

**Comment by karol-bisztyga** [comment]
I also took a look at the docs and it doesn't seem like they need to be updated, but I might've missed something. In the CLI reference, there are no mentions for other commands like `declare` or `deploy` of what they return. In other docs where the `build` command is mentioned, the output doesn't seem to be used and there are no cases for this.

I was also thinking about this:

> consider printing the class hash optional 

and I think we can leave this as is, which means the class hash will always be printed out. I think this works the same for `declare` and `deploy`.

I'm going to treat the current state of this PR as a final one. Again, if you disagree, feel free to share your opinion. Thanks.

File: tests/e2e/test_building.py [hunk 5]
```diff
@@ -9,11 +10,26 @@
...
+    output = protostar(["build"])
+    assert 'Class hash for contract "main": ' in output
+    assert output.split('Class hash for contract "main": ')[1].strip()[:2] == "0x"
+    numeric_value = int(
+        output.split('Class hash for contract "main": ')[1].strip()[2:], 16
+    )
+
+    output_json = protostar(["build", "--json"])
+    output_json_parsed = json.loads(output_json.split("\n")[0])
+    assert output_json_parsed["main"][:2] == "0x"
```
Comment (MaksymilianDemitraszek) [inline tests/e2e/test_building.py:29]: [[Nitpick]]
```suggestion
    assert output_json_parsed["main"].startswith() == "0x"
```

**Review by MaksymilianDemitraszek — APPROVED** [review]

File: tests/e2e/test_building.py (hunk 5 above)
Comment (karol-bisztyga) [reply tests/e2e/test_building.py]: [use startswith() as suggested](https://github.com/software-mansion/protostar/pull/1244/commits/fa36790dacd35d306616ea97242206e62e3a0070)

**Pull request merged**